docs/rfc/0662-collaborative-objects.adoc

```diff
@@ -0,0 +1,690 @@
+})
+const change = Automerge.getChanges(doc, updatedDoc)
+const changeBytes = uint8ToBase64(change)
+----
+
+What we do here is load the automerge document from it's history, then use the
+automerge Javscript library to mutate the document (the `Automerge.change`
+call) and then finally get the change between the original version of the 
+document and the new one. 
+
+Now that we have the change we can make a `PATCH` request to 
+`<radicle-implementation>/projects/<project URN>/metadata/<metadata ID>` with
+the following contents:
+
+[source,json]
+----
+{
+    "changes": {
+        "type": "automerge",
+        "change": "<some base64>"
+    }
+}
+----
+
+This endpoint will return an error if the change does not match the schema of
+the object. Otherwise the change will be merged in to the object and announced
+to the network.
+
+== Implementation
+
+Every collaborative object is represented by a graph of automerge changes. See
+the <<appendix_automerge>> for more about automerge. We map this change graph
+to git as described in <<Change Commits>>, we store the tip of the change 
+graph under a `/cob/<typename>/<object ID>` reference, this allows us to 
+reconstruct the state of a collaborative object for a particular peer as 
+described in <<Reconstructing Collaborative Objects>>. This allows us to use
+git to replicate collaborative objects, see <<Fetching Collaborative Objects>>.
+
+We require some schema extensions to authenticate data in collaborative objects,
+these are described in <<Schema Extensions>>.
+
+
+=== Change Commits
+
+Given that automerge changes are a hash linked graph, we can map them to Git.
+We do so by wrapping each change in a commit. The commit points at a tree with
+the following layout
+
+[source]
+----
+.
+|--change
+|--manifest.toml
+----
+
+This tree contains a batch of automerge changes to a collaborative object. The
+changes must all be from a single actor ID, which is the peer ID of the peer
+which created the change. We will go into more details shortly. Any direct
+dependencies of this change are encoded in the same manner and become the
+parents of this commit. This allows us to reconstruct the automerge depdency
+graph. 
+
+Along with the dependencies of the commit we also need to add the commit of the
+identity which created this commit. We need this identity to validate
+signatures and by making the commit a parent we ensure that git will replicate
+it for us. 
+
+A valid change commit must have three trailers:
+
+* `X-Rad-Signature`, as for identity documents
+* `X-Rad-Author-Parent`, this is the hash of the commit which references the
+  author identity. We use this trailer to avoid following the author commit
+  reference when constructing the automerge change graph
+* `X-Rad-Schema-Parent`, this is the hash of the parent commit which contains 
+  the schema of this object. See [schema commits](#schema-commits).
+
+Furthermore a change commit MUST be signed by an identity which was at some point
```

Comment:
I wonder if this limitation is required. Perhaps we could use a similar model of how branches are handled in identities. That is, all peers can add changes to their view of the object, but it is the maintainers' view that is the canonical view (or most important view). So in essence, non-maintainers are proposing comments to an issue, for example.

Reply:
This does have some appealing characteristics. Mainly that it is simple to
implement and it builds on intuitions people already need to have about Radicle.

However, I am also not sure that this is a good solution when we start thinking
about what collaborative objects are for. The canonical branch solution works
because maintainers are actively maintining those branches. Specifically they
are very aware of the meaning of the difference between different branches. When
a non-maintainer proposes a branch the maintainer will look at what the diff of
the new branch is and decide whether to track it as a canonical branch.

The situation with collaborative objects is different. The reason we want to
constrain objects to maintainers only is that there are assertions we want to
make about a collaborative object without having to examine the underlying data
structure. Consider the case of attributing authorship on an issue comment. This
RFC proposes no mechanism for ensuring that the comment was actually authored by
the person who's URN is on the comment. This means that in order to accept a
proposed issue a maintainer would need to examing the before and after state of
the entire issue checking that each comments author URN is valid. This is in
some sense the same work as examining a proposed branch, except that now it's
for a data structure the user has no understanding of or tooling for. 

I think this approach _could_ work but we would need to provide tooling for
maintainers to examine the difference between states of collaborative objects
and ways for schema authors to indicate what parts of an object should be
checked. This might be worth investigating if the completely automated
authorization system I envision turns out not to be possible.

This approach also feels promising for schema migrations.

Reply:
Well regarding authorship, if the author points to a URN, this has a set of keys associated with it. Now, if there's also a signature trailer, one can verify the signature from the set of keys. Would this work for verification?

Also, that's specific to authorship so is there any other kind of authorization we would need to plan for?

Reply:
The issue is allowing collaborative object designers to express the
authorization constraints in a way which is general enough to accomodate any
automerge document. We could hard code an authorization rule about issue comment
URNs, but then we are doing something very application specific. We would end up
writing a rule something like "if a change modifies a comment (identified by a
path into the JSON document) then ensure that the change is signed by the user
who made the comment and that the URN on the comment is the URN of the user
making the change".

There are a bunch of such rules you might want for issues, but very different
rules for things like kanban boards, and we don't want any of those rules to be
in `radicle-link`.

This is the work I want to address in the next step, which is a sort of
combination of authorization and schema validation (these being the same problem
in a decentralised CRDT). I'm currently thinking that we include a datalog with
constraints based authorization DSL which is part of the first commit of the
object and might look something like this:

```
creator says that X can insert into <comments>
where X is a user

creator says that Y.urn must be urn(X)
where Y isChild(<comments>)
and X isAuthor(Y)
```

This is some kind of logic over paths and URNs, which currently is very fuzzy
but I think it's doable. It's inspired by https://ieeexplore.ieee.org/abstract/document/4271637